src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,444 @@
+            + ".HAS_MEMBERS",
+        "com.puppycrawl.tools.checkstyle.checks.coding.AvoidDoubleBraceInitializationCheck"
+            + ".IGNORED_TYPES",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck"
+            + ".ALLOWED_ASSIGNMENT_CONTEXT",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck"
+            + ".ALLOWED_ASSIGNMENT_IN_COMPARISON_CONTEXT",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck.COMPARISON_TYPES",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck.CONTROL_CONTEXT",
+        "com.puppycrawl.tools.checkstyle.checks.coding.InnerAssignmentCheck"
+            + ".LOOP_IDIOM_IGNORED_PARENTS",
+        "com.puppycrawl.tools.checkstyle.checks.coding.MatchXpathCheck.xpathExpression",
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.AtclauseOrderCheck.DEFAULT_ORDER",
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocBlockTagLocationCheck.DEFAULT_TAGS",
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.SummaryJavadocCheck.ALLOWED_TYPES",
+        "com.puppycrawl.tools.checkstyle.checks.modifier.ModifierOrderCheck.JLS_ORDER",
+        "com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck"
+            + ".TOKENS_FOR_INTERFACE_MODIFIERS",
+        "com.puppycrawl.tools.checkstyle.checks.regexp.RegexpMultilineCheck.detector",
+        "com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineCheck.detector",
+        "com.puppycrawl.tools.checkstyle.checks.coding.IllegalTokenTextCheck.formatString",
+        "com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck.tagRegExp",
+        "com.puppycrawl.tools.checkstyle.checks.naming.AbstractNameCheck.format",
+        "com.puppycrawl.tools.checkstyle.checks.whitespace.AbstractParenPadCheck.option"
+    );
+
+    /**
+     * List of fields that are a zero size array. They are immutable by definition.
+     */
+    private static final Set<String> ZERO_SIZE_ARRAY_FIELDS = Set.of(
+        "com.puppycrawl.tools.checkstyle.utils.CommonUtil.EMPTY_BIT_SET",
+        "com.puppycrawl.tools.checkstyle.utils.CommonUtil.EMPTY_BYTE_ARRAY",
+        "com.puppycrawl.tools.checkstyle.utils.CommonUtil.EMPTY_DOUBLE_ARRAY",
+        "com.puppycrawl.tools.checkstyle.utils.CommonUtil.EMPTY_INTEGER_OBJECT_ARRAY",
+        "com.puppycrawl.tools.checkstyle.utils.CommonUtil.EMPTY_INT_ARRAY",
+        "com.puppycrawl.tools.checkstyle.utils.CommonUtil.EMPTY_OBJECT_ARRAY",
+        "com.puppycrawl.tools.checkstyle.utils.CommonUtil.EMPTY_STRING_ARRAY"
+    );
+
+    /**
+     * List of classes not following {@code classesWithImmutableFieldsShouldBeStateless} and
+     * {@code classesWithMutableFieldsShouldBeStateful} ArchRule.
+     */
+    private static final Set<String> SUPPRESSED_CLASSES = Set.of(
+        // Failing 'testClassesWithImmutableFieldsAreStateless'
```

Comment:
Separate lists will eliminate the need for these comments, otherwise we would need to update them to new name of rule.

Reply:
Created separate lists.